source/particle/property/cpo_bingham_average.cc

```diff
@@ -0,0 +1,369 @@
+            sum_matrix_b[2][2] += matrices[i_grain][1][2] * matrices[i_grain][1][2]; // SUM(n^2)
+            sum_matrix_b[0][1] += matrices[i_grain][1][0] * matrices[i_grain][1][1]; // SUM(l*m)
+            sum_matrix_b[0][2] += matrices[i_grain][1][0] * matrices[i_grain][1][2]; // SUM(l*n)
+            sum_matrix_b[1][2] += matrices[i_grain][1][1] * matrices[i_grain][1][2]; // SUM(m*n)
+
+
+            sum_matrix_c[0][0] += matrices[i_grain][2][0] * matrices[i_grain][2][0]; // SUM(l^2)
+            sum_matrix_c[1][1] += matrices[i_grain][2][1] * matrices[i_grain][2][1]; // SUM(m^2)
+            sum_matrix_c[2][2] += matrices[i_grain][2][2] * matrices[i_grain][2][2]; // SUM(n^2)
+            sum_matrix_c[0][1] += matrices[i_grain][2][0] * matrices[i_grain][2][1]; // SUM(l*m)
+            sum_matrix_c[0][2] += matrices[i_grain][2][0] * matrices[i_grain][2][2]; // SUM(l*n)
+            sum_matrix_c[1][2] += matrices[i_grain][2][1] * matrices[i_grain][2][2]; // SUM(m*n)
+
+          }
+        const std::array<std::pair<double,Tensor<1,3,double>>, 3> eigenvectors_a = eigenvectors(sum_matrix_a, SymmetricTensorEigenvectorMethod::jacobi);
+        const std::array<std::pair<double,Tensor<1,3,double>>, 3> eigenvectors_b = eigenvectors(sum_matrix_b, SymmetricTensorEigenvectorMethod::jacobi);
+        const std::array<std::pair<double,Tensor<1,3,double>>, 3> eigenvectors_c = eigenvectors(sum_matrix_c, SymmetricTensorEigenvectorMethod::jacobi);
+
+
+        const Tensor<1,3,double> averaged_a = eigenvectors_a[0].second * eigenvectors_a[0].first;
+        const Tensor<1,3,double> averaged_b = eigenvectors_b[0].second * eigenvectors_b[0].first;
+        const Tensor<1,3,double> averaged_c = eigenvectors_c[0].second * eigenvectors_a[0].first;
+
+        return
+        {
+          {
+            {{averaged_a[0],averaged_a[1],averaged_a[2]}},
+            {{averaged_b[0],averaged_b[1],averaged_b[2]}},
+            {{averaged_c[0],averaged_c[1],averaged_c[2]}}
+          }
+        };
+      }
+
+      template<int dim>
+      std::vector<Tensor<2,3>>
+      CpoBinghamAverage<dim>::random_draw_volume_weighting(std::vector<double> fv,
+                                                           std::vector<Tensor<2,3>> matrices,
+                                                           unsigned int n_output_grains) const
+      {
+        // Get volume weighted euler angles, using random draws to convert odf
+        // to a discrete number of orientations, weighted by volume
+        // 1a. Get index that would sort volume fractions AND
+        //ix = np.argsort(fv[q,:]);
+        // 1b. Get the sorted volume and angle arrays
+        std::vector<double> fv_to_sort = fv;
+        std::vector<double> fv_sorted = fv;
+        std::vector<Tensor<2,3>> matrices_sorted = matrices;
+
+        unsigned int n_grain = fv_to_sort.size();
+
+
+        /**
+         * ...
+         */
```

Comment:
This needs documentation. Also your algorithm here is N^2 in the size of the vectors (linear because of your loop and linear because of max_element) and quite expensive per step, because you sort both vectors. What is the typical size of these vectors? If they are bigger than 100 I would consider a better algorithm, e.g. the one suggested in this answer on stackoverflow: https://stackoverflow.com/a/17074810 (you can replace the `compare` template with <). Also you do not really need to sort the matrices, you just need to keep a permutation from the sorted to the unsorted array (which this other algorithm gives you as byproduct). That way you can keep the vector of matrices untouched.